snuba/optimize.py

```diff
@@ -226,6 +319,12 @@ def optimize_partitions(
         query = (query_template % args).strip()
         logger.info(f"Optimizing partition: {part.name}")
         start = time.time()
+
+        # Update tracker before running clickhouse.execute since its possible that the connection can get
+        # disconnected while the server is still executing the optimization. In that case we don't want to
+        # run optimization on the same partition twice.
+        if optimize_partition_tracker:
+            optimize_partition_tracker.update_completed_partitions(part.name)
```

Comment:
I don't think this conceptually correct. 
In this implementation, every time we restart the job without the state being cleaned `get_partitions_to_optimize` is re-executed and the partitions that were already processed are excluded.
The problem is that `get_partitions_to_optimize` already pre-filters partitions that do not need to be optimized. Granted, most of the times we may need to optimize everything but can happen.
So, without keeping in redis the whole list of partitions you wanted to optimize on a given day, you may end up optimizing a partition after a restart (let's say at 5 am) that was not in the list for that day right after midnight because no merge had touched it that day. 
This will make you optimize more partitions during a day than we are doing today. 

I would store the state slightly differently: 
- at the beginning of the job, store (with the current date) the list of partitions to optimize.
- every time a partition is completed flag it as complete
- if you restart during the same day, just load the state and do not ask which partitions you need to Clickhouse and trust the state in redis.
At the start also clean up any possible remaining state from the previous date (but you can just set a TTL for the entire set to 24 hours).

This also ensures that the state starts fresh every day.

Reply:
> Moreover, in the current state, if the job restarts and there is a merge on going (merges are asynchronous, if the job crashes, the job does not stop) it will not care about on going merges and just schedule a new one.
This is a problem and we should avoid it. Imagine if K8S is having a bad day and keeps restarting this job over and over again. We will keep scheduling new merges. Clickhosue does not respect any quota on OPTIMIZE FINAL calls. So we may take down clickhouse.

>I think it is important to keep the state of the last merge we scheduled and query system.merges when we start. If we find that a merge we had started but not completed is still on going, just wait fora moment when clickhouse will be doing no merges.

This PR has already gotten pretty big than I would have liked it to be. I have incorporated all suggestions except the one mentioned above. I would like to tackle that in a separate PR to keep things sane. In the meantime, we could avoid doing parallel clickhouse optimizations even if this PR is merged.